### ISSUE TITLE
`torch.nn.GELU` cannot be serialized within a dataclass but can otherwise be serialized

### ISSUE THREAD

**Issue by tom-hehir** [omni-us/jsonargparse#739]
## 🐛 Bug report

When I try to serialize a dataclass containing `torch.nn.GELU` I am warned that jsonargparse is "Unable to serialize instance GELU(approximate='none')" and the corresponding field in the serialized output is populated with this warning message. If I try to serialize `torch.nn.GELU` outside of a dataclass, it behaves as expected and produces a useful serialized representation.

### To reproduce

Code:
```python

import torch.nn as nn
from jsonargparse import ArgumentParser
from jsonargparse import lazy_instance
from dataclasses import dataclass


@dataclass
class MyDataclass:
    activation: nn.Module = nn.GELU()


class MyClass: ...


@dataclass
class MyPseudoDataclass(MyClass):
    activation: nn.Module = nn.GELU()


def parse_gelu():
    parser = ArgumentParser()
    parser.add_argument(
        "--activation",
        type=nn.Module,
        default=lazy_instance(nn.GELU, approximate="none"),
    )
    args = parser.parse_args([])
    json_str = parser.dump(args)
    return json_str

def parse_dataclass():
    parser = ArgumentParser()
    parser.add_argument(
        "--config",
        type=MyDataclass,
        default=lazy_instance(MyDataclass)
    )
    args = parser.parse_args([])
    json_str = parser.dump(args)
    return json_str

def parse_pseudo_dataclass():
    parser = ArgumentParser()
    parser.add_argument(
        "--config",
        type=MyPseudoDataclass,
        default=lazy_instance(MyPseudoDataclass)
    )
    args = parser.parse_args([])
    json_str = parser.dump(args)
    return json_str


if __name__ == '__main__':
    print("Serialized GELU output:")
    print(parse_gelu())
    print("Serialized MyDataclass output:")
    print(parse_dataclass())
    print("Serialized MyPseudoDataclass output:")
    print(parse_pseudo_dataclass())

```

Output:
```
Serialized GELU output:
activation:
  class_path: torch.nn.GELU
  init_args:
    approximate: none

Serialized MyDataclass output:
config:
  activation: Unable to serialize instance GELU(approximate='none')

Serialized MyPseudoDataclass output:
config:
  class_path: __main__.MyPseudoDataclass
  init_args:
    activation: Unable to serialize instance GELU(approximate='none')

JsonargparseWarning: Unable to serialize instance GELU(approximate='none')
```

### Expected behavior

In each of the 3 scenarios above, `activation` should be populated with:
```
class_path: torch.nn.GELU
init_args:
  approximate: none
```

### Environment

- jsonargparse version: 4.40.0
- Python version: 3.12.8
- How jsonargparse was installed: pip
- OS: macOS

**Comment by tom-hehir** [comment]
Note: I just realized that if I wrap the GELU as a lazy instance in the dataclass definitions then I obtain the expected behaviour. I am unsure if this is the intended behaviour, but it was unintuitive to me personally.

Many thanks for the great package -- I couldn't do my work without it!

**Comment by mauvilsa** [comment]
The behavior you are observing is the currently expected one, and not a bug. Have a read at [default-values](https://jsonargparse.readthedocs.io/en/v4.40.0/#default-values).

Note what the problem is. When serializing (i.e. figuring out correct `class_path` and `init_args`), the code just gets access to an instance of a class. In general in python it is not possible to know how a class instance was created, which is why you see that warning. It is simply all that can be done.

Also note that using class instances as defaults is a bad practice. The reason being that if the class is used multiple times, then the exact same default would be used in every place. This can lead to difficult to catch bugs. Not a problem for `torch.nn.GELU`, but would be bad for an `torch.nn.Module` with trainable parameters.

The use of `lazy_instance` I agree is not great since it is jsonargparse-specific. But there are other possibilities to avoid class instance defaults. One possibility could be for classes to receive functions that instantiate the class. For the activation example above could be to have instead `activation_fn: Callable[[], nn.Module] = nn.GELU`. Also see for example optimizers in [callable-type](https://jsonargparse.readthedocs.io/en/v4.40.0/#callable-type). This obviously requires more changes in the code, because the functions need to be called at some point to create the class instance.

**Comment by tom-hehir** [comment]
Ah thank you for clarifying this and the detailed advice -- 100% makes sense!

I hadn't considered the issue with setting `nn.Module` instances as defaults but will definitely be changing my code to avoid this (luckily, it looks like I escaped doing it with modules that have learnable params). I'm also already using the callable style defaults in some other parts of my code and quite like this approach so will consider switching to that!

Closing issue with this comment :)
